### DISCUSSION TITLE
Rendering pages based on the role

### DISCUSSION THREAD

**Discussion by Hadi-bakhshi** [vercel/next.js#37336] [Help]
Hello,
I'm new to NextJS, and I want to implement a dashboard which will have one layout, different roles, different pages and content base on the role. I came up with an idea but I have no clue whether it's possible. The idea is that the user enter the URL I want to send only the login page to him not anything else. When that guy logs in, based on that person's role I want to send the pages that I created in the page folder for that role:
please, check out this picture 

![Untitled](https://user-images.githubusercontent.com/85115222/171168555-66e5aeb6-33cc-4453-a4af-e2f046a58871.png)

I would be appreciated if you guide me
Thanks

**Comment by ephraimduncan** [comment]
I have an idea, you can try it to see if it works. 

You can use a Next.js `middleware` for this. 

- In the middleware, you check for the role of the user
- Redirect to a specific route based on the role.

Here's how I would implement it, you can modify it for yourself.
```js
// _middleware,js
import { NextResponse } from "next/server";

export function middleware(req) {
  const url = req.url;

  if (url.includes("/login")) {
    // get the role here
    const role = "ADMIN";

    // Check and redirect to pages here
    if (role === "ADMIN") {
      NextResponse.redirect("/admin");
    } else if (role === "SUPERVISOR") {
      NextResponse.redirect("/supervisor");
    } else {
      NextResponse.redirect("/user");
    }
  }

  return NextResponse.next();
}

```

**Reply by Hadi-bakhshi** [reply]
Thanks for your answer. I have a question, I don't want to send the pages that are not for them, I mean if I'm an admin and I sign in, the pages for supervisor doesn't exist for me. Is it possible with this approach?
Thanks in advance

**Reply by ephraimduncan** [reply]
Yes. You can further implement on per page basis in the middleware to reroute the page if the role is not allowed. Something like this

```js
if (url.includes("/supervisor")) {
    if (role !== "SUPERVISOR") {
      NextResponse.redirect("/login");
    }
}
```

You can redirect to the root route or login page. 

If you found this helpful, please mark this as an answer.
